lte/gateway/c/session_manager/LocalEnforcer.cpp

```diff
@@ -587,24 +590,72 @@ void LocalEnforcer::complete_redirect_flow_install(
     MLOG(MDEBUG) << "Session for IMSI " << imsi << " not found";
     return;
   }
-  auto session_update = session_store_.get_default_session_update(session_map);
 
-  // check if the rule has been installed already.
+  RuleLifetime lifetime{};
+  auto session_update = session_store_.get_default_session_update(session_map);
+  auto& uc = session_update[imsi][session_id];
   for (const auto& session : it->second) {
-    if (session->get_session_id() == session_id &&
-        !session->is_dynamic_rule_installed(rule.id())) {
-      MLOG(MDEBUG) << "Install redirect GY flow in pipelined for "
-                   << session_id;
-      pipelined_client_->add_gy_final_action_flow(
-          imsi, ip, static_rules, {rule});
-
-      auto& uc = session_update[imsi][session_id];
-      session->insert_gy_dynamic_rule(rule, lifetime, uc);
+    if (session->get_session_id() == session_id) {
+      switch (final_action_info.action_type) {
+        case REDIRECT:
+          {
+            // This is GY based REDIRECT, GX redirect will come in as a regular rule
+            std::vector<std::string> static_rules;
+            const auto& rule = create_redirect_rule(final_action_info);
+            // check if the rule has been installed already.
+            if (!session->is_gy_dynamic_rule_installed(rule.id())) {
```

Comment:
let me ask again this, because I may be missing something: why REDIRECT case is checking for the rule session->is_gy_dynamic_rule_installed and for Restrict we are not checking if the restrict_rules are already installed?

Reply:
The existing check for is meaningless.
first, redirect rule is inserted regardless of pipelined returned value as It is async call and thus doesn't guarantee pipelined installation.
second, we have two states redirected and restricted that guarantee the same (not creating another redirect action)

1283 if (grant->service_state == SERVICE_REDIRECTED) {
1284 MLOG(MDEBUG) << "Redirection already activated for " << session_id_ ;
1285 continue;
1286 }